pkg/envoy/ads/profile.go

```diff
@@ -45,3 +49,34 @@ func (s *Server) trackXDSLog(cn certificate.CommonName, typeURL envoy.TypeURI) {
 		s.xdsLog[cn][typeURL] = timeSlice
 	})
 }
+
+// validateRequestResponse is a utility function to validate the response generated by a given request.
+// currently, it checks that all resources responded for a request are being responded to,
+// or will log as `warn` otherwise
+// Returns the number of resources NOT being answered for a request. For wildcard case, this is always 0.
+func validateRequestResponse(proxy *envoy.Proxy, request *xds_discovery.DiscoveryRequest, respResources []types.Resource) int {
+	// No validation is done for wildcard cases
+	resourcesRequested := mapset.NewSet()
+	resourcesToSend := mapset.NewSet()
+
+	// Get resources being requested
+	for _, reqResource := range request.ResourceNames {
+		resourcesRequested.Add(reqResource)
+	}
+
+	// Get resources being responded by name
+	for _, res := range respResources {
+		resourcesToSend.Add(cache.GetResourceName(res))
+	}
+
+	// Compute difference from request resources' perspective
+	resDifference := resourcesRequested.Difference(resourcesToSend)
+	diffCardinality := resDifference.Cardinality()
+	if diffCardinality != 0 {
+		log.Warn().Msgf("Proxy SerialNumber=%s PodUID=%s: not all request resources for type %s are being responded to req [%v] resp [%v] diff [%v]",
```

Comment:
Is it possible due to timing scenarios where a requested resource is not responded to because the requested resource is no longer relevant as computed by the controller?

Reply:
yes, though if the resource is no longer available/relevant, that should get reflected also on some next update which will/should remove all references to it from all verticals (this is an assumption given our way of doing things).

So, in general, this case is possible but I'd say unlikely. I suggested/committed it as `warn` precisely because it does not automatically imply error, but it's very relevant information to know when debugging imo. wdyt?

Reply:
Thanks for clarifying, that makes sense.